go/types/struct.go

```diff
@@ -16,46 +16,58 @@ import (
 )
 
 var EmptyStructType = MakeStructType("")
-var EmptyStruct = Struct{ValueSlice{}, EmptyStructType, &hash.Hash{}}
+var EmptyStruct = Struct{"", structValueFields{}, EmptyStructType, &hash.Hash{}}
 
 type StructData map[string]Value
 
+type structValueField struct {
+	name  string
+	value Value
+}
+
+type structValueFields []structValueField
+
+func (fs structValueFields) Len() int           { return len(fs) }
+func (fs structValueFields) Swap(i, j int)      { fs[i], fs[j] = fs[j], fs[i] }
+func (fs structValueFields) Less(i, j int) bool { return fs[i].name < fs[j].name }
+
 type Struct struct {
-	values []Value
+	name   string
+	fields structValueFields
```

Comment:
I think it'd be nice to keep values as a `ValueSlice` and just add `name String` and `fieldNames []string`. Having values just be a value slice was motivated by wanting a fast path to just hand in a slice of values and the constructor wouldn't need to do any more allocation.

In the new world, I think maybe the "fast-path" constructor should hand in name, fieldNames and valueSlice. the constructor would need to validate that len(fieldNames) == len(valueSlice) and that fieldNames are sorted and valid idents.